src/ray/object_manager/object_manager.cc

```diff
@@ -409,6 +410,59 @@ void ObjectManager::PushLocalObject(const ObjectID &object_id, const NodeID &nod
                      std::move(release_chunk_callback));
 }
 
+void ObjectManager::PushSpilledObject(const ObjectID &object_id, const NodeID &node_id,
+                                      const std::string &spilled_url) {
+  // SpilledObject::CreateSpilledObject does synchronous IO; schedule it off
+  // main thread.
+  rpc_service_.post(
+      [this, object_id, node_id, spilled_url, chunk_size = config_.object_chunk_size]() {
+        auto optional_spilled_object =
+            SpilledObject::CreateSpilledObject(spilled_url, chunk_size);
+        if (!optional_spilled_object.has_value()) {
+          RAY_LOG(WARNING) << "Failed to load splled object " << object_id
+                           << ". It may have been evicted.";
+          return;
+        }
+
+        auto spilled_object =
+            std::make_shared<SpilledObject>(std::move(optional_spilled_object.value()));
+
+        uint64_t total_data_size =
+            spilled_object->GetDataSize() + spilled_object->GetMetadataSize();
+        uint64_t metadata_size = spilled_object->GetMetadataSize();
+        uint64_t num_chunks = spilled_object->GetNumChunks();
+        rpc::Address owner_address = spilled_object->GetOwnerAddress();
+
+        auto spilled_object_chunk_reader = [object_id, spilled_object](
+                                               uint64_t chunk_index,
+                                               rpc::PushRequest &push_request) -> Status {
+          auto optional_chunk = spilled_object->GetChunk(chunk_index);
+          if (!optional_chunk.has_value()) {
+            RAY_LOG(WARNING) << "Read chunk " << chunk_index << " of object " << object_id
+                             << " failed. "
+                             << " It may have been evicted.";
+            return Status::IOError("Failed to read spilled object");
```

Comment:
I think here has the same concern as Eric's comment above. When is the time this function fails vs the above function `CreateSpilledObject ?

Reply:
It might be pretty hard to propagate this error properly. We'd need to surface it as a app-level exception, which is doable but requires changing the size / type of the object to an error message.

Maybe for now we can leave it as LOG(ERROR) and add a TODO. It's tempting to RAY_CHECK here but I think this is too risky given the likelihood of race conditions from stale push requests even after the object is deleted.

Reply:
Yup that's my conclusion too after reading the code. To handle it correct it might require a bit plumbing the error around. Currently added a test to demonstrate it hangs..